- Improved package organization

matuki opened the pull request:
- Implemented "apply wallpaper" feature, adding respective states

Commit:
- Improved package organization
- Implemented "apply wallpaper" feature, adding respective states

igorescodro commented on app/src/main/java/com/pmatuki/wowpapers/application/WowpapersApplication.kt:
```diff
@@ -1,11 +1,14 @@
+import javax.inject.Inject
 
-class WowpapersApplication: Application() {
+class WowpapersApplication @Inject constructor() : Application() {
```
You probably shouldn't inject the constructor of Application class once you won't inject it anywhere else. I don't know exactly how ToothPick works, but probably you need to have a function here that pass this class as a parameter to the application context be added in the graph, not annotating it.

matuki replied:
Done!

igorescodro commented on app/src/main/java/com/pmatuki/wowpapers/view/detail/DetailActivity.kt:
```diff
@@ -58,10 +60,23 @@ class DetailActivity : AppCompatActivity() {
                     wallpaperDrawable = state.imageDrawable
                     this@DetailActivity.binding.backgroundImage.setImageDrawable(wallpaperDrawable)
                 }
+                DetailViewState.Applying -> {
```
You mentioned that you want me to review the states from this Activity and does look good to me. I prefer to have a more complex state handling, with multiple options like you did here, instead of having more than one state. In my personal experience, always prefer to have a Single Source of Truth, like you did here.

One way of maybe make the code simpler for the UI (specially if you will have more states here) is creating a `data class` containing all the view properties from your screen. You would build that component there on the ViewModel and only pass the class to be rendered here.

For example:
```
data class DetailViewState(
    val loadingVisibility: Int,
    val cardVisibility: Int,
    val applyButtonEnabled: Boolean
)
```
Then your ViewModel will build this object based on the states and the Activity will only do something like:
```
fun setupState(state: DetailViewState) {
    layoutViewLoading.visibility = state.loadingVisibility
    cardView.visibility = state.cardVisibility
    applyButton.isEnabled = state.applyButtonEnabled
}
```

However, at the moment I prefer the implementation you did here. I don't like to give too much knowledge about the view to the ViewModel. But in very complex screens, this is an alternative.

igorescodro commented on app/src/main/java/com/pmatuki/wowpapers/view/detail/DetailViewModel.kt:
```diff
@@ -16,6 +19,9 @@ internal class DetailViewModel : ViewModel() {
     @Inject
     lateinit var imageDownloadService: ImageDownloadService
 
+    @Inject
+    lateinit var wallpaperService: WallpaperService
+
```
This injections should be added in the ViewModel constructor. If not, it will be very difficult to test or reuse this ViewModel if needed.

matuki replied:
Done!

Commit:
- Removed unnecessary DI on DetailViewModel and WallpaperListViewModel
- Added ViewModelFactory implementation to allow passing constructor
parameters to ViewModels.

matuki commented on app/src/main/java/com/pmatuki/wowpapers/view/list/ListActivity.kt:
```diff
@@ -29,7 +33,12 @@ class ListActivity : AppCompatActivity(), WallpaperItemClickListener {
-        viewModel = ViewModelProvider(this).get(WallpaperListViewModel::class.java)
+        viewModel = ViewModelProvider(
+            this, WallpaperListViewModelFactory(
+                WallpaperDataSource(WallhavenService()), WallpaperMapper()
```
I changed ViewModels and I'm passing dependencies in the constructor now.

However, I don't like that ListActivity has to know so much about what dependencies are needed to build a WallpaperDataSource. It has to reference WallhavenService and WallpaperMapper, which feels wrong.

@igorescodro : should I inject the dependencies inside the Factory class instead? After all, shouldn't a factory class  know how to build the object it is meant to provide without the Activity (or any other "client") having to specify its dependencies?

igorescodro replied:
Yes, you are right about feeling that this approach is not correct.
This here basically breaks the principle where a more-complex component needs to know how to instantiate a less-complex component (here we need to create instances of Service and Mapper to be able to use the ViewModel).

The ViewModel creation should be in your Module class or directly annotated on the ViewModel itself (`@Inject`)

I believe that this sample could help you better understand it:
https://github.com/villela/toothpickviewmodelexample
